### DISCUSSION TITLE
Eask is reporting no dependencies when using global workspace

### DISCUSSION THREAD

**Discussion by mohkale** [emacs-eask/cli#176] [Q&A] [answered]
Hi there,

I'm just trying to get started. I've created the following Eask file at `Eask` in the root of my project.

```elisp
;; -*- emacs-lisp -*-

(package "projection"
         "0.1.0"
         "Project type support for project.el")

(website-url "https://github.com/mohkale/projection")
(keywords "project" "convenience")

(package-file "src/projection.el")
(files "src/*.el" "src/projection-multi/*.el")

(source "gnu")
(source "melpa")

(depends-on "emacs" "29")
(depends-on "compile-multi")

(development
 (depends-on "f")
 (depends-on "buttercup"))
```

When I run eask install it fails.

```text
root@822d3d1ae205:/workarea# eask install-deps -g
(No dependencies found in your Eask file)
''                                                                ''

    <?> You can add dependencies by using specifier [depends-on]

        [+] (depends-on "PKG-SPEC")

''                                                                ''
```

Did I miss something in the docs? I'm not sure why it's failing :thinking:.

**Comment by jcs090218** [comment, marked as answer]
~~In development scope, you need to add flag `--dev`:~~

```sh
eask install-deps -g --dev
```

---

Ah, okay. I didn't correctly read the question the first time.

With `-g` flag, your elpa directory is pointed to `~/.eask/`, so you need your Eask-file to be there and not the project root.

**Reply by mohkale** [reply]
Thanks this fixed the issue. Now when I run install-deps it succeeds.

```text
Unmatched dependency 'compile-multi'; add (compile-multi "VERSION") to package-file or consider removing it
Installing 3 package dependencies...

  - [1/3] Skipping project (0)... already installed X
  - [2/3] Skipping compat (29.1.4.1)... already installed X
  - [3/3] Skipping compile-multi (20230311.1951)... already installed X

(Total of 0 dependencies installed, 3 skipped)

Installing 2 development dependencies...

  - [1/2] Skipping f (20230116.1032)... already installed X
  - [2/2] Skipping buttercup (20230418.819)... already installed X

(Total of 0 dependencies installed, 2 skipped)
```

**Comment by mohkale** [comment]
Hope you won't mind me using this discussion to ask other debugging questions. I tried running `eask -g emacs -nw` and then `M-: (require 'compat)` but that's failing with an error that the package is not installed. How can I run emacs within the eask environment and all the package accessible?

**Reply by jcs090218** [reply]
> Hope you won't mind me using this discussion to ask other debugging questions.

It might be helpful if you opened another discussion so other people can see it. :)

>  tried running eask -g emacs -nw and then M-: (require 'compat) but that's failing with an error that the package is not installed. How can I run emacs within the eask environment and all the package accessible?

I tried it and it should work 🤔

```
eask install compat -g
eask -g emacs
```

in scratch buffer:

```
(require 'compat)

(message "%s" (featurep 'compat))  ; t
```

**Reply by mohkale** [reply]
@jcs090218 

> It might be helpful if you opened another discussion so other people can see it.

Good point. I'll create a new one for subsequent questions.

> I tried it and it should work

Hmm... Still failing for me :disappointed:. Here's a reproducible recipe with Docker in case it helps.

Create a test.Dockerfile file containing the following:

```dockerfile
FROM silex/emacs:master-ci

COPY Eask /root/.eask/Eask
RUN apt-get update \
 && apt-get install -y unzip \
 && curl -fsSL https://raw.githubusercontent.com/emacs-eask/cli/master/webinstall/install.sh | sh
ENV PATH="$PATH:/root/.local/bin"
```

Create an Eask file in the same directory containing the following:

```lisp
;; -*- emacs-lisp -*-

(package "projection"
         "0.1.0"
         "Project type support for project.el")

(website-url "https://github.com/mohkale/projection")
(keywords "project" "convenience")

(package-file "src/projection.el")
(files "src/*.el" "src/projection-multi/*.el")

(source "gnu")
(source "melpa")

(depends-on "emacs" "29")
(depends-on "project" "0.9.8")
(depends-on "compat" "29.1.4.1")

;; Extension packages
(depends-on "compile-multi")

(development
 (depends-on "f")
 (depends-on "buttercup"))
```

Now build the image and start a shell.

```bash
docker build --tag=test-eask --file=test.Dockerfile .
docker run --rm -it test-eask bash
```

Repeat the same test you just shared and it fails with compat cannot be found :/.

**Reply by jcs090218** [reply]
I'm confused. Are you trying to mix `local` and `global` scopes? Your Eask-file looks like a "local" (regularly for elisp project) one.

For local scope (without -g):

```sh
eask install-deps --dev
eask -g emacs                 # This wouldn't work
```

For global scope, you need to copy your Eask-file to `~/.eask/` (make it a global Eask-file) and:

```sh
eask install-deps --dev -g
eask -g emacs                   # This will work.
```

Better to create a global Eask-file instead of copying the one you had.

**Reply by mohkale** [reply]
> Are you trying to mix local and global scopes?

Not intentionally, I'm not quite sure what the difference is. In this case I'm just copying my local projects Eask file to the global location and running install-deps afterwards. I would've figured if I supply the `-g` flag then the one I have in my local repo directory wouldn't matter. It may help if you take a look at this [commit](https://github.com/mohkale/projection/commit/95e4555db96ff7a43f409ef39f3b6ae0abebbd67) which shows what I was working with.

Is there something that's supposed to be different in the contents of a global or package file?

**Reply by jcs090218** [reply]
The biggest difference is where to load your packages. If you use -g, only the global packages will be enabled for you! The local workspace is for the project-only sandbox. Therefore, local and global are two isolated sandboxes.

**Reply by mohkale** [reply]
> The biggest difference is where to load your packages. If you use -g, only the global packages will be enabled for you

That makes sense, in this case I'm supplying `-g` so is there a reason the compat package isn't being loaded even thought it's installed :thinking:.

**Reply by jcs090218** [reply]
It seems like it's installed in the local one, and that's why it's not found because eask is looking at the compat package in the global scope.

I am going to make a graph to explain this; please stay tuned.

**Reply by jcs090218** [reply]
I've added an image to show how Eask works under the hood. See https://emacs-eask.github.io/Getting-Started/Basic-Usage/#-knowing-your-elpa-directory.
